src/Server/TCPHandler.cpp

```diff
@@ -1285,37 +1267,57 @@ void TCPHandler::receiveQuery()
 
+    /// TODO unify interserver authentication (currently this code looks like a backdoor at a first glance)
     if (is_interserver_mode)
     {
+        client_info.interface = ClientInfo::Interface::TCP_INTERSERVER;
 #if USE_SSL
+        String user_for_session_log = client_info.initial_user;
+        if (user_for_session_log.empty())
+            user_for_session_log = USER_INTERSERVER_MARKER;
+        String cluster_secret = server.context()->getCluster(cluster)->getSecret();
+        if (salt.empty() || cluster_secret.empty())
+        {
+            auto exception = Exception(ErrorCodes::AUTHENTICATION_FAILED, "Interserver authentication failed");
+            session->onAuthenticationFailure(AlwaysAllowCredentials{USER_INTERSERVER_MARKER}, socket().peerAddress(), exception);
+            throw exception; /// NOLINT
+        }
+
         std::string data(salt);
         data += cluster_secret;
         data += state.query;
         data += state.query_id;
         data += client_info.initial_user;
 
-        if (received_hash.size() != 32)
-            throw NetException("Unexpected hash received from client", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
-
         std::string calculated_hash = encodeSHA256(data);
+        assert(calculated_hash.size() == 32);
 
+        /// TODO maybe also check that peer address actually belongs to the cluster?
         if (calculated_hash != received_hash)
-            throw NetException("Hash mismatch", ErrorCodes::UNEXPECTED_PACKET_FROM_CLIENT);
-        /// TODO: change error code?
+        {
+            auto exception = Exception(ErrorCodes::AUTHENTICATION_FAILED, "Interserver authentication failed");
+            session->onAuthenticationFailure(AlwaysAllowCredentials{USER_INTERSERVER_MARKER}, socket().peerAddress(), exception);
+            throw exception; /// NOLINT
+        }
 
+        /// NOTE Usually we get some fields of client_info (including initial_address and initial_user) from user input,
+        /// so we should not rely on that. However, in this particular case we got client_info from other clickhouse-server, so it's ok.
         if (client_info.initial_user.empty())
         {
             LOG_DEBUG(log, "User (no user, interserver mode)");
+            session->authenticateInterserverFake();
         }
         else
         {
             LOG_DEBUG(log, "User (initial, interserver mode): {}", client_info.initial_user);
             session->authenticate(AlwaysAllowCredentials{client_info.initial_user}, client_info.initial_address);
         }
 #else
-        throw Exception(
+        auto exception = Exception(
             "Inter-server secret support is disabled, because ClickHouse was built without SSL library",
-            ErrorCodes::SUPPORT_IS_DISABLED);
+            ErrorCodes::AUTHENTICATION_FAILED);
+        session->onAuthenticationFailure(AlwaysAllowCredentials{USER_INTERSERVER_MARKER}, socket().peerAddress(), exception);
+        throw exception; /// NOLINT
```

Comment:
Is there any benefits from this style? I mean why are you writing

```
auto exception = Exception(...);
throw exception;
```

instead of just

```
throw Exception(...)
```

Reply:
Because this exception is also passed to `onAuthenticationFailure`, so actually it's
```
auto exception = Exception(...);
someExtraLogging(exception);
throw exception;
```